base/threadfuncs.jl

```diff
@@ -0,0 +1,54 @@
+# This file is a part of Julia. License is MIT: https://julialang.org/license
+
+"""
+    Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads(),
+                    schedule=:fair)
+
+Similar to `foreach(f, channel)`, but iteration over `channel` and calls to
+`f` are split across `ntasks` tasks spawned by `Threads.@spawn`. This function
+will wait for all internally spawned tasks to complete before returning.
+
+If `schedule` is `:fair`, `Threads.foreach` will spawn a new task for each work
+item, allowing Julia's scheduler to more freely load-balance work items across
+threads. Technically, in `:fair` mode, `Threads.foreach` may execute up to
+`2 * ntasks` tasks concurrently.
+
+If `schedule` is `:static`, `Threads.foreach` will simply spawn `ntasks` tasks
+such that each task is locked to the thread it was spawned on. This approach has
+lower scheduling overhead and thus may be more suitable for more uniform,
+fine-grained workloads.
+
+!!! compat "Julia 1.6"
+    This function requires Julia 1.6 or later.
+"""
+function Threads.foreach(f, channel::Channel; ntasks=Threads.nthreads(),
+                         schedule=:fair)
+    if schedule === :fair
+        apply = _waitspawn
+    elseif schedule === :static
+        apply = (f, x) -> f(x)
+    else
+        throw(ArgumentError("`schedule` must be either `:fair` or `:static`"))
+    end
+    stop = Threads.Atomic{Bool}(false)
+    @sync for _ in 1:ntasks
+        Threads.@spawn try
```

Comment:
Yet another note for future improvement: I think, in principle, we can use `jl_set_task_tid` here. However, it'd freeze at `@sync` if there are long-running task on some threads without yields. Leaking a task for such a case could be an option but it would pile up in the task queue so I don't think it's a good solution. I think we need a task cancellation API for this, too.